### DISCUSSION TITLE
Applying a ControlTheme to a UserControl in code

### DISCUSSION THREAD

**Discussion by LaurentInSeattle** [AvaloniaUI/Avalonia#15791] [Q&A]
Hi all, 

I am trying to replicate / recycle for Avalonia some old WPF code that was applying (WPF) Style objects to (WPF) UserControl.

So first, I created a ControlTheme in XAML: 
`
	<UserControl.Resources>
		<ControlTheme x:Key="GlyphButtonTypography" TargetType="TextBlock">
			<Setter Property="Background" Value="Yellow"/>
			<Setter Property="Foreground" Value="Red"/>
			<Setter Property="LetterSpacing" Value="1.5"/>
			<Setter Property="FontSize" Value="12"/>
			<Setter Property="FontWeight" Value="SemiBold"/>
			<Setter Property="FontStyle" Value="Italic"/>
		</ControlTheme>
	</UserControl.Resources>
`
Then created a StyledProperty for my custom UserControl: 

`    
    public static readonly StyledProperty<ControlTheme> TypographyProperty =
        AvaloniaProperty.Register<GlyphButton, ControlTheme>(
            nameof(Typography),
            defaultValue: new ControlTheme(),
            inherits: false,
            defaultBindingMode: BindingMode.OneWay,
            validate: null,
            coerce: null,
            enableDataValidation: false);

    public ControlTheme Typography
    {
        get => this.GetValue(TypographyProperty);
        set
        {
            this.SetValue(TypographyProperty, value);

            // Does not work :( 
            this.textBlock.Theme = value;

            bool applied = this.textBlock.ApplyStyling();  // Returns true, but does nothing
            this.textBlock.InvalidateVisual();             // Does not help 

            // Wild West style 
            var mi = typeof(StyledElement).GetMethod(
                "ApplyControlTheme", BindingFlags.Instance | BindingFlags.NonPublic, new Type[] { });
            if (mi != null)
            {
                mi.Invoke(this.textBlock, new object[] { }); // Does nothing
                this.textBlock.InvalidateVisual();            // Does not help 
            }

            // Work around 
            foreach (var item in value.Setters)
            {
                var setter = item as Setter;
                if ((setter is not null) && (setter.Property is not null))
                {
                    // Works, except for FontStyle property
                    this.textBlock.SetCurrentValue(setter.Property, setter.Value);
                }
            }
        }
    }
`
Assigning the ControlTheme to the Theme property was expected to work, but does not.
I tried some other stuff and finally applied the properties "by hand", that worked.

Is it the way it is supposed to work? 
Am I doing anything wrong ? 
Or is there a better approach than passing down ControlTheme objects as Avalonia properties ? 

On a side note, FontStyle Italic does not work. Using Font Inter and Fluent as the default style on Win10. 
(not tested on other platforms)

**Comment by stevemonaco** [comment]
It seems like the major problem here is a `TextBlock` is not a `TemplatedControl`. It's not a lookless control therefore there's no `ControlTheme` to define its visual appearance -- it renders itself (or via a helper), similar to `Border`.

It seems like you should define a `<Style Selector="TextBlock.someClass">` for this, assign `Classes="someClass"`, and call it a day. You can define+add `Style`s in code, too.

**Reply by LaurentInSeattle** [reply]
The doc's say: 

Control themes are typically applied to [templated (lookless)](https://docs.avaloniaui.net/docs/basics/user-interface/controls/creating-controls/choosing-a-custom-control-type) controls, but they can actually be applied to any control. 

So, it should work on TextBlock. And it does in XAML but not in code.

**Comment by stevemonaco** [comment]
For the font question, I'm assuming you're using the default `Inter` as provided by Avalonia. AFAIK, Avalonia doesn't ship the italic variants, only some weights. https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Fonts.Inter/Assets

I might be missing something though.

**Reply by LaurentInSeattle** [reply]
Regarding the font: Yes, italics are not shipping, which is a bit confusing because the Italic FontStyle is mentioned many times as an example in the docs.
I have downloaded Inter-Variable from here: https://www.fontsquirrel.com/fonts/inter 
There is one TTF in the archive with all styles, this "Inter-V.TTF" is quite small and covers all styles.
Not sure about the quality of the typography though...

**Reply by stevemonaco** [reply]
~~Avalonia currently doesn't support variable fonts yet, IIRC.~~

**Reply by LaurentInSeattle** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/79057037/975c72a6-1013-4a06-b805-ac68b25abb68)

Works here...

**Reply by stevemonaco** [reply]
Ah, this works on nightly (and probably earlier), but not 11.0.10 for me. I wasn't aware that this was already added.

**Reply by LaurentInSeattle** [reply]
I'm using the 11.0.10 nugets from nuget.org dated Tuesday, March 5, 2024, but only tested on Windows

**Reply by stevemonaco** [reply]
Maybe something was messed up in my testing because I've downgraded the test project to 11.0.10 and now it works there, too. 🤷‍♂️
